### ISSUE TITLE
Confirmation on the Reproduction Process of Results

### ISSUE THREAD

**Issue by 2814894078** [shiyao-li/NextBestPath#8]
Dear Authors,  

Currently, I am working on reproducing the experimental results for the **Normal difficulty level** of the AiMDoom dataset presented in Table 2. To avoid discrepancies caused by potential misunderstandings, I would like to confirm the accuracy of the core reproduction process with you.  

Based on the descriptions in the paper, I have outlined the steps for reproducing the results of the Normal difficulty level as follows:  
1. **Training Phase**: Run `python train_nbp.py` to train the NBP model and save the weight file with the strongest generalization ability (AiMDoom_normal_best_val.pth).  
2. **Testing Phase**: Load the aforementioned `AiMDoom_normal_best_val.pth` weights and run `python test_nbp_planning.py` to test each of the 30 test scenes under the path `NBP/AiMDoom_Normal/test_dataset` one by one. Obtain the `Final Coverage` and `AUCs` metrics for each scene.  
3. **Result Statistics**: Calculate the "mean ± standard deviation" for the `Final Coverage` and `AUCs` across all 30 test scenes. The final result should match the values for "NBP (Ours)" in the Normal difficulty level of Table 2 (i.e., `0.734±0.142` for Final Coverage and `0.526±0.112` for AUCs).  

I would greatly appreciate your confirmation on whether the above process fully aligns with your team’s experimental design. Additionally, could you please clarify how many epochs are used for training, or roughly at which epoch the best validation weights (e.g., AiMDoom_normal_best_val.pth) are typically obtained? Could you clarify if there are any critical details I may have missed?  

Thank you for your time and assistance—your clarification is crucial for me to accurately reproduce the experimental results.

**Comment by shiyao-li** [comment]
Nice to hear your feedback again. Your understanding of the reproduction process is correct.

Additionally, here is a useful tip to make your testing phase more efficient:
You do not need to manually modify the config files to add each new test scene from time to time. Instead, you can simply read all scene names under `AiMDoom_Normal/test_dataset` and load them all at once, either directly into the dataloader or by adding all of them into the config file in a single step.

Regarding the number of training epochs, this is quite flexible. Since our training pipeline uses curriculum learning and each training episode starts from a random scene and a random initial pose, the total number of epochs is not strictly fixed. This randomness leads to some variation in the training process. Based on our experiments, training on a single A6000 GPU for approximately one day (as mentioned in the paper) is sufficient to reach strong performance. You may adjust the total training time according to your own GPU and computational capability.

**Comment by 2814894078** [comment]
Dear Author,

Thank you for your patient responses. With your guidance, I successfully trained the model for the **Hard segment** and conducted testing. Following the method we discussed, I evaluated the model on 30 test scenarios (each containing 5 trajectories) and calculated the average value. However, there is a noticeable gap between my result and the one in Table 2 (0.31 vs. 0.61). 

<img width="873" height="725" alt="Image" src="https://github.com/user-attachments/assets/e04845cf-6c0e-4214-bce4-51a661a851ff" />


I suspected issues with my training process, so I downloaded the AiMDoom_hard_best_val.pth file you provided and re-ran the test—but the result remained unsatisfactory.


<img width="900" height="714" alt="Image" src="https://github.com/user-attachments/assets/f73bce07-7e26-492c-8606-92af4fb1b4eb" />


I wonder if the problem lies in my calculation method. Taking the file appalling_river_10_results.json as an example: for the first trajectory, the "coverage" data is a time-series list (as shown below). Currently, I am using the last value in this list (e.g., 0.32741427421569824 for this trajectory) as the "Final Coverage" for that trajectory, then averaging the Final Coverage values of all 5 trajectories in a scenario to get the scenario’s average. Could you confirm if this calculation method is correct?

Additionally, I would greatly appreciate it if you could explain how to calculate the AUC metric. Since your project does not include code for metric calculation, I have been calculating it manually, which has been quite challenging. Thank you for your help!


<img width="1067" height="785" alt="Image" src="https://github.com/user-attachments/assets/d7fb6490-74da-4613-b1bd-085f7b3baa09" />

**Comment by shiyao-li** [comment]
Hi, I am happy to help. 

When reproducing the results in the paper, please double-check the following:
1. When training in the Hard-level scenes, the test scenes should also be in Hard-level.
2. The maximum allowed steps differ across difficulty levels — Simple, Normal, Hard, Insane — with limits of 100, 200, 400, and 500, respectively. You have to change [this](https://github.com/shiyao-li/NextBestPath/blob/main/next_best_path/testers/nbp_planning.py#L60).

From your results, it seems you didn’t update the maximum step limit.

For computing the AUC and Final Coverage of each trajectory, you could use the following code:
```
coverage_list = coverage_list[:max_steps]
dx = 1 / max_steps
def compute_auc(coverage_list, dx):
    auc = np.trapz(coverage_list, dx=dx) + coverage_list[0] * dx / 2.
    return auc

# Final coverage is the maximum value in the trajectory
max_cov = max(coverage_list)
```

**Comment by 2814894078** [comment]
Thank you for your reply. I have now solved this problem and the code is currently running in the testing phase. At the same time, I would also like to ask: if it is necessary to set the maximum step during testing, is it also required during training? If yes, where should I set it?

**Comment by shiyao-li** [comment]
Adjusting the maximum step limit in the testing phase according to the scene difficulty is essential, because the size and layout complexity vary across difficulty levels. 

However, this adjustment is not necessary in the training phase.

We mentioned these details in the experimental section of the paper.

**Comment by 2814894078** [comment]
I deeply appreciate your patient and professional assistance on multiple occasions! Each revision has precisely addressed my needs, allowing me to successfully reproduce the experimental results. Please accept my sincere gratitude once again!

**Comment by 2814894078** [comment]
Dear Authors,

I'm writing to seek your assistance regarding some discrepancies I encountered while testing your method. Using the pre-trained weights provided in your README (https://drive.google.com/drive/folders/1jAEKrznbbZ5bwu39y0ah4pszMlTuVAfH), I tested the Simple and Normal scenarios, but the results I obtained differ significantly from those reported in Table 2. I would greatly appreciate it if you could help me identify the potential issues.

<img width="1789" height="569" alt="Image" src="https://github.com/user-attachments/assets/429400e2-8a2f-4395-8820-68b3a1a41943" />

 

To provide context, I have confirmed the following key details:

1. The number of steps was set strictly according to your specifications: 100 steps for the Simple scenario and 200 steps for the Normal scenario.
2. I used the pre-trained weights directly as provided in your shared drive without any modifications.

I suspect the inconsistency may stem from my evaluation methodology. For your reference, I have attached the code used for result evaluation and the corresponding output files (.json format). Could you kindly review these materials to check for any errors in my implementation?

[author_result.zip](https://github.com/user-attachments/files/23879413/author_result.zip)

Additionally, I observed an unexpected behavior during testing related to memory storage. For example, in the directory：

.AiMDoom_Simple/test_dataset/back_to_the_oppressive_sandstorm_9/test_memory_0/
├── training/0/frames  
├── training/1/frames  
├── training/2/frames

├── training/3/frames  
└── training/4/frames

Only the `training/0/frames` folder contains `.pt` files, while the other subdirectories (1-4) are empty. Could this empty memory storage be a contributing factor to the inconsistent results compared to Table 2?

Thank you very much for your time and assistance. I look forward to your valuable feedback.

**Comment by shiyao-li** [comment]
Hi, I'm wondering whether, during data collection, the training data from the Simple and Normal scenes might have been mixed into the same LMDB, which could lead to degraded performance.

In the table, the value before “±” represents the mean, and the value after it represents the standard deviation. So your results are actually not that bad, please note that the standard deviation in the table is quite large. The reason for this large variance is that the complexity of scenes within the same difficulty level is not perfectly consistent, and it’s difficult to make them fully consistent. If you run multiple tests, in my experience, the results tend to be roughly half better and half worse compared to the results in the table.

Regarding your observation that only `training/0/frames` contains `.pt `files: this behavior is expected, or more precisely, intentional. Writing everything into a single folder reduces inode usage on the server, and the additional folders were created solely to assist with debugging.

**Comment by 2814894078** [comment]
Thank you for your consistent and patient explanations all along. This has resolved my doubts, and I wish you every success in your research!

**Comment by shiyao-li** [comment]
Thanks! All the best with your research too.
